doc/port_auto_neg/port-auto-negotiation-design.md

```diff
@@ -0,0 +1,452 @@
+
+#### DB Migrator Enhancements
+
+In current SONiC implementation, if auto negotiation is enabled, it uses the `speed` field as the advertised speeds. Since this feature introduced a new field `adv_speeds`, we need do DB migration to keep backward compatible. For example, the configuration: 
+
+```json
+"Ethernet0": {
+    ...
+    "autoneg": "1",
+    "speed": "100000"
+}
+```
+
+Will be migrated to:
+
+```json
+"Ethernet0": {
+    ...
+    "autoneg": "1",
+    "speed": "100000",
+    "adv_speeds": "100000"
+}
+```
+
+#### SWSS Enhancements
+
+The current SONiC speed setting flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if autoneg == true and speed is set:
+    setPortAdvSpeed(port, speed)
+else if autoneg == false and speed is set:
+    setPortSpeed(port, speed)
+```
+
+The new SONiC speed setting flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if autoneg == true:
+    speed_list = vector()
+    if adv_speeds is set:
+        speed_list = adv_speeds
+    else:
+        speed_list = all_supported_speeds
+    setPortAdvSpeed(port, speed_list)
+
+    interface_type_list = vector()
+    if adv_interface_types is set:
+        interface_type_list = adv_interface_types
+    else:
+        interface_type_list = all_supported_interface_types
+    setPortAdvInterfaceType(port, interface_type_list)
+else:
+    if speed is set:
+        setPortSpeed(port, speed)
+    if interface_type is set:
+        setInterfaceType(port, interface_type)
+```
+
+#### portsyncd and portmgrd Consideration
+
+No changes for portsyncd and portmgrd.
+
+Due to historical reason, portsyncd and portmgrd both handle **PORT** table changes in **CONFIG_DB** and write **APPL_DB** according to configuration change. portmgrd handles fields including "mtu", "admin_status" and "learn_mode"; portsyncd handles the rest fields. 
+
+#### Port Breakout Consideration
+
+No changes for port breakout.
+
+Currently, when user change port breakout mode, SONiC removes the old ports and create new ports in CONFIG_DB according to platform.json. For example, say Ethernet0 supports breakout mode 1x100G and 2x50G. The platform.json looks like:
+
+```json
+"interfaces": {
+	"Ethernet0": {
+	    "index": "1,1,1,1",
+	    "lanes": "0,1,2,3",
+	    "alias_at_lanes": "Eth1/1, Eth1/2, Eth1/3, Eth1/4",
+	    "breakout_modes": "1x100G,2x50G"
+	 },
+ 	 ...
+ }
+```
+
+**Without port auto negotiation feature**,  the current configuration of Ethernet0 is:
+
+```json
+"Ethernet0": {
+    "alias": "Eth1/1",
+    "lanes": "0,1,2,3",
+    "speed": "100000",
+    "index": "1"
+}
+```
+
+User change the port breakout mode via command `config interface breakout Ethernet0 2x50G`. According to platform.json, the configuration of Ethernet0 will change to:
+
+```json
+"Ethernet0": {
```

Comment:
autoneg is default disabled, but FW is doing autoneg enabled. what could be the affect for port breakout.

Reply:
This is not an issue since SAI and SONiC side will keep backward compatible. Based on that, leaving port breakout unchanged will also keep backward compatible.